agents/check_mk_agent.freebsd

```diff
@@ -581,11 +580,23 @@ run_purely_synchronous_sections() {
         echo '<<<statgrab_net>>>'
         statgrab net. 2>&1 | cut -d. -f2-99 | sed 's/ *= */ /'
 
-        echo '<<<statgrab_mem>>>'
-        grep -E "^(swap|mem)\." "${tmpfile}" | sed 's/ *= */ /'
-
         rm -f "${tmpfile}"
     fi
+
```

Comment:
I figured it would be easier to stay compatible with the existing way statgrab works to keep complexity to a minimum.  Are you sure you'd rather it done this way?

Reply:
Hello,
I would like to help about this patch. 
I agree with the idea of `<<<sysctl_mem>>>` but there is some tools that relies on `<<<statgrab_mem>>>` and have "difficulties" to be updated to `<<<sysctl_mem>>>` if this section exists only on FreeBSD platform, which you know is less deployed than others.
How can I help?

Reply:
Oh, no! I completely lost track of this :-( . I am so sorry!

I'm not sure if I understand your question. Generally, it should never be a problem if a section is not present in the output. As far as I can see, `<<<statgrab_mem>>>` is only mentioned in some places that are a relic from the time there was a dedicated plugin.

Maybe I should elaborate: These days (>2.0), we have one additional layer of abstraction: The difference between raw and parsed sections. For example: The [check plugin "mem_used"](https://github.com/Checkmk/checkmk/blob/f62a9bcb6059eb6658d692b505a84ada804a018b/cmk/base/plugins/agent_based/mem_used.py#L201). It doesn't have an explicit `sections` argument, so it subscribes to exactly one section, by the same name as itself by default. That means, whenever there is a *parsed* section by the name `mem_used`, this plugin will be called.

Looking at it from the agent's end: The agent creates a *raw* section with the name `statgrab_mem`. The server receives this output, and picks the agent [section plugin by that name](https://github.com/Checkmk/checkmk/blob/46d93451de0fb10e92fa9f3266a430d847e12c73/cmk/base/plugins/agent_based/mem_used_sections.py#L279). It now applies the parse function, and stores the result under the *parsed* section name.

Now we have a parsed section called "mem_used", and the plugin will be called.

In this instance we already have several sections that create a "mem_used" section in the end ("aix_memory", "solaris_mem", "statgrab_mem", "openbsd_mem"). The check plugin in end does not need to know where the parsed section originated from -- it will just do its thing.

The advantage of this approach is that you move a lot of the parsing logic into the python world on server side, which makes it much easier to test.

Reply:
Hello.
Sorry for the late reply, I have looked about `check_mk` code. And I feel a bit puzzled.
On FreeBSD, the package check_mk_agent contains (mostly) only `agents/check_mk_agent.freebsd` and that's all. No plugins are installed or ported.
So I understand better the idea of @brd for this patch.
In a nutshell, do we need to install plugins to have memory information ?

Reply:
That would be an option, but no, that is not what I meant. It is always a bit of a judgement call what goes directly into the agent, and what should be a plug-in, but I think memory is something that should unquestionably be monitored out of the box.

Im afraid it is quite confusing with all the different things we refer to as "plug-in". In this case only few are relevant (but need to be distinguished):
 * "agent plug-ins": plugins to the agent, deployed on the monitored host, found in `agents/plugins`. Not wanted here.
 * "agent section plugin": a plug-in to the check- and discovery engine, run on the monitoring server, found in `cmk/base/plugins/agent_based` or - preferably - `cmk/plugins/<FAMILY>/agent_based`. This is what we need.
 * "check plug-in": also a plug-in to the check- and discovery engine, run on the server, found in same folder. By using the "parsed_section_name" feature we don't have to write it.

There are two files that need to be adjusted here I think:
 1) The agent file (and only this needs to be deployed on the monitored host): I think the agent should try to create the new section `sysctl_mem`, and only if that fails fall back to `statgrab_mem` (a construct that can be seen [here](https://github.com/Checkmk/checkmk/blob/4890ed4c61464e0f2cfaf0f496ccde96171b80b3/agents/check_mk_agent.linux#L2190), but we don't need the condition and the logging).
 2) The agent section plugins as linked above. This file does not need to be deployed to the monitored host, it is run by the server.

If you want, you could start by sending something unfinished as PR, so that I can try to guide you through the process via review step by step. This will require some back and forth, obviously.

Reply:
I think that makes sense.